In [1]:
import pandas as pd
import os
import pandas as pd
import numpy as np
from scipy import stats
import logging

In [2]:
import os, sys
project_root = os.path.abspath(os.path.join(os.getcwd(), '..'))
if project_root not in sys.path:
    sys.path.append(project_root)

In [3]:
from src.hypothesis_testing.hypothesis_testing import InsuranceHypothesisTester, run_pipeline_orchestrator

In [4]:
df=pd.read_csv('../data/processed/cleaned_insurance_data.csv')

In [5]:
hypo_insurance=run_pipeline_orchestrator(df)
hypo_insurance


 EXECUTIVE HYPOTHESIS TESTING STATISTICAL REPORT (RECONSTRUCTED) 

HYPOTHESIS 1: Regional Risk Differences (Claim Frequency)
  Target Segments : Control: Western Cape vs. Test: Gauteng
  Control Metric  : 0.2093% Claim Frequency
  Test Metric     : 0.3359% Claim Frequency
  Calculated p-val: 1.376276e-15
  Outcome         : REJECT H0
----------------------------------------------------------------------
HYPOTHESIS 2: Geographic Risk Differences (Claim Severity)
  Target Segments : Control: Postal Code 2000 vs. Test: Postal Code 122
  Control Metric  : Average Severity of 19,196.41
  Test Metric     : Average Severity of 18,162.03
  Calculated p-val: 7.002080e-01
  Outcome         : FAIL TO REJECT H0
----------------------------------------------------------------------
HYPOTHESIS 3: Geographic Performance Differences (Net Margin)
  Target Segments : Control: Postal Code 2000 vs. Test: Postal Code 122
  Control Metric  : Average Net Margin of -8.16
  Test Metric     : Average Net Margi


### Executive Hypothesis Testing Report

#### 1. Summary of Decisions

* **Hypothesis 1 (Province vs. Claim Frequency):** **REJECT $H_0$**. Location significantly impacts how often claims are made.
* **Hypothesis 2 (Zip Code vs. Claim Severity):** **FAIL TO REJECT $H_0$**. Zip codes do not change the average cost of a claim.
* **Hypothesis 3 (Zip Code vs. Net Margin):** **FAIL TO REJECT $H_0$**. There is no significant profit difference between the compared zip codes.
* **Hypothesis 4 (Gender vs. Claim Frequency):** **REJECT $H_0$**. Men claim significantly more often than women.



#### 2. Findings & Business Interpretations

##### Hypothesis 1: Regional Risk (Claim Frequency)

* **What the data shows:** The claim frequency in the Western Cape is **0.2093%**, while Gauteng is significantly higher at **0.3359%** ($p = 1.38 \times 10^{-15}$).
* **Observation:** Drivers in Gauteng are roughly **60% more likely** to file a claim than drivers in the Western Cape. This is a massive, real risk difference.
* **Business Move:** Stop charging the same base rate across the country. Charge higher base premiums for vehicles operating in Gauteng to cover the higher accident rates.

##### Hypothesis 2: Geographic Risk (Claim Severity)

* **What the data shows:** The average claim cost in Postal Code 2000 is **19,196.41**, compared to **18,162.03** in Postal Code 122 ($p = 0.700$).
* **Observation:** The ~1,000 unit difference is completely due to random statistical noise. Once an accident happens, the cost to repair or replace the vehicle is structurally the same in both areas.
* **Business Move:** Do not split your claims or repair budgets by specific zip codes. Treat urban accident costs as uniform.

##### Hypothesis 3: Geographic Performance (Net Margin)

* **What the data shows:** The average net margin per policy is **-8.16** for Postal Code 2000 and **-22.86** for Postal Code 122 ($p = 0.246$).
* **Observation:** Both of your highest-volume zip codes are actively losing money. The statistical test proves that neither zip code is performing significantly better than the other—they are both trapped in an underperforming deficit.
* **Business Move:** This is not a localized problem with one "bad" zip code. It is a systemic pricing issue. You need a blanket premium increase across high-volume urban policies to push the margins back into the positive.

##### Hypothesis 4: Demographic Risk (Gender Claim Frequency)

* **What the data shows:** By using customer titles to bypass the missing data, we analyzed 99.92% of the portfolio. Reconstructed women have a claim frequency of **0.1978%**, while reconstructed men sit at **0.2833%** ($p = 6.65 \times 10^{-5}$).
* **Observation:** Men are **43% more likely** to file a claim than women. Because this was tested across almost the entire dataset, the gap is definitive.
* **Business Move:** Gender is a clear risk driver for how often a claim occurs. Incorporate this frequency gap into your segmentation strategy to price male drivers accurately according to their higher likelihood of filing claims.


#### Hypothesis Testing Results Summary Table

| Hypothesis | Metric / KPI | Statistical Test Used | $p$-value | Decision |
| --- | --- | --- | --- | --- |
| **$H_0$: No risk differences across provinces.** | Claim Frequency | Chi-Squared ($\chi^2$) Test of Independence | $1.38 \times 10^{-15}$ | **REJECT $H_0$** |
| **$H_0$: No risk differences between zip codes.** | Claim Severity |  Two-Sample $t$-test | $7.00 \times 10^{-1}$ | **FAIL TO REJECT $H_0$** |
| **$H_0$: No significant margin difference between zip codes.** | Net Margin | Two-Sample $t$-test | $2.46 \times 10^{-1}$ | **FAIL TO REJECT $H_0$** |
| **$H_0$: No significant risk difference between Women and Men.** | Claim Frequency | Chi-Squared ($\chi^2$) Test of Independence | $6.65 \times 10^{-5}$ | **REJECT $H_0$** |



#### Business Recommendations for Rejected Hypotheses

##### 1. Regional Pricing Adjustments (Based on Province Rejection)

* **Finding:** Drivers in Gauteng are roughly **60% more likely** to file a claim than drivers in the Western Cape.
* **Recommendation:** Charging a single flat base rate across different regions is causing underpricing in high-risk zones. Stop using uniform national pricing. Introduce a mandatory regional risk-loading factor (premium surcharge) for all assets registered or operated in Gauteng to pull that segment back to profitability, while keeping the Western Cape as a lower-premium baseline to remain competitive.

##### 2. Risk-Adjusted Demographic Segments (Based on Gender Reconstructed Proxy Rejection)

* **Finding:** By using customer titles to circumvent the corrupted data, the full portfolio revealed that men are **43% more likely** to file a claim than women.
* **Recommendation:** Gender is a definitive driver of accident *frequency*. Because the likelihood of an asset filing a claim changes structurally based on this demographic, you must integrate this frequency gap directly into your upcoming risk segmentation model. Factor this higher probability into the premium calculations for male drivers, or utilize high-integrity proxy fields (like vehicle performance tiers or occupation types) to protect the model's target underwriting margins.